File: connexus_app/ios/Runner/AudioHandler.swift
import Foundation
import AVFoundation
import Flutter

class AudioHandler: NSObject {
  private let audioSession = AVAudioSession.sharedInstance()

  func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "setAudioMode":
      if
        let args = call.arguments as? [String: Any],
        let mode = args["mode"] as? String
      {
        setAudioMode(mode: mode, result: result)
      } else {
        result(
          FlutterError(
            code: "INVALID_ARGS",
            message: "Missing mode argument",
            details: nil
          )
        )
      }

    case "setSpeaker":
      if
        let args = call.arguments as? [String: Any],
        let enabled = args["enabled"] as? Bool
      {
        setSpeaker(enabled: enabled, result: result)
      } else {
        result(
          FlutterError(
            code: "INVALID_ARGS",
            message: "Missing enabled argument",
            details: nil
          )
        )
      }

    case "setMute":
      if
        let args = call.arguments as? [String: Any],
        let muted = args["muted"] as? Bool
      {
        setMute(muted: muted, result: result)
      } else {
        result(
          FlutterError(
            code: "INVALID_ARGS",
            message: "Missing muted argument",
            details: nil
          )
        )
      }

    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func setAudioMode(mode: String, result: @escaping FlutterResult) {
    do {
      switch mode {
      case "voice_call":
        try audioSession.setCategory(
          .playAndRecord,
          mode: .voiceChat,
          options: [.allowBluetooth, .defaultToSpeaker]
        )
        try audioSession.setActive(true)
      case "normal":
        try audioSession.setActive(false)
        try audioSession.setCategory(.ambient)
      default:
        break
      }
      result(true)
    } catch {
      result(
        FlutterError(
          code: "AUDIO_ERROR",
          message: error.localizedDescription,
          details: nil
        )
      )
    }
  }

  private func setSpeaker(enabled: Bool, result: @escaping FlutterResult) {
    do {
      if enabled {
        try audioSession.overrideOutputAudioPort(.speaker)
      } else {
        try audioSession.overrideOutputAudioPort(.none)
      }
      result(true)
    } catch {
      result(
        FlutterError(
          code: "SPEAKER_ERROR",
          message: error.localizedDescription,
          details: nil
        )
      )
    }
  }

  private func setMute(muted: Bool, result: @escaping FlutterResult) {
    // iOS does not expose direct microphone mute; this is expected to be
    // handled at the WebRTC / Telnyx SDK level. We simply report success.
    result(true)
  }
}

File: connexus_app/android/app/src/main/kotlin/com/example/connexus_app/AudioHandler.kt
package com.example.connexus_app

import android.content.Context
import android.media.AudioManager
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class AudioHandler(private val context: Context) : MethodChannel.MethodCallHandler {

    private val audioManager: AudioManager by lazy {
        context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "setAudioMode" -> {
                val mode = call.argument<String>("mode")
                setAudioMode(mode, result)
            }

            "setSpeaker" -> {
                val enabled = call.argument<Boolean>("enabled") ?: false
                setSpeaker(enabled, result)
            }

            "setMute" -> {
                val muted = call.argument<Boolean>("muted") ?: false
                setMute(muted, result)
            }

            else -> result.notImplemented()
        }
    }

    private fun setAudioMode(mode: String?, result: MethodChannel.Result) {
        try {
            when (mode) {
                "voice_call" -> {
                    audioManager.mode = AudioManager.MODE_IN_COMMUNICATION
                    audioManager.isSpeakerphoneOn = false
                }

                "normal" -> {
                    audioManager.mode = AudioManager.MODE_NORMAL
                    audioManager.isSpeakerphoneOn = false
                }
            }
            result.success(true)
        } catch (e: Exception) {
            result.error("AUDIO_ERROR", e.message, null)
        }
    }

    private fun setSpeaker(enabled: Boolean, result: MethodChannel.Result) {
        try {
            audioManager.isSpeakerphoneOn = enabled
            result.success(true)
        } catch (e: Exception) {
            result.error("SPEAKER_ERROR", e.message, null)
        }
    }

    private fun setMute(muted: Boolean, result: MethodChannel.Result) {
        try {
            audioManager.isMicrophoneMute = muted
            result.success(true)
        } catch (e: Exception) {
            result.error("MUTE_ERROR", e.message, null)
        }
    }
}


